File: lab/8/mandel/mandel.cpp
#include <boost/python/module.hpp>
#include <boost/python/def.hpp>
#include <boost/python/list.hpp>
#include <complex>
#include <iostream>

namespace py = boost::python;

inline std::complex<double> pixel_to_point(
    size_t width, size_t height,
    size_t x, size_t y,
    std::complex<double> upper_left,
    std::complex<double> lower_right
) {
    double range_x = lower_right.real() - upper_left.real();
    double range_y = upper_left.imag() - lower_right.imag();
    return std::complex<double>{
        upper_left.real() + double(x) * range_x / double(width),
        upper_left.imag() - double(y) * range_y / double(height)
    };
}

namespace mandelbrot {
    typedef std::complex<double> complex;

    unsigned char escapes(const complex &c, const unsigned &limit, const unsigned empty) {
        auto z = complex{0, 0};
        for (unsigned i = 0; i < limit; i++) {
            z = z*z + c;
            if (z.real()*z.real() + z.imag()*z.imag() >= 4) {
                return i;
            }
        }

        return empty;
    }

    py::list render(size_t width, size_t height, const complex &ul, const complex&lr) {
        py::list list;
        for (size_t y = 0; y < height; y++) {
            for (size_t x = 0; x < width; x++) {
                auto c = pixel_to_point(width, height, y, x, ul, lr);

                list.append(255 - mandelbrot::escapes(c, 255, 255));
            }
        }
        return list;
    }
}

BOOST_PYTHON_MODULE(mandel) {
    py::def("escapes", mandelbrot::escapes);
    py::def("render", mandelbrot::render);
}

File: lab/8/mandel.py
"""Lab 7: Mandelbrot Set part 2

USAGE: mandel.py <width> <height> <upper-left> <lower-right>

Options:
    -h, --help    Show this screen
    --version     Show the version number

Print the mandelbrot set to screen using a C++ utility to generate the data
faster than a native python application can.
"""

from docopt import docopt
import numpy as np
from pylab import imshow, show
import time
import mandel


def pixel_to_point(width, height, x, y, ul, lr):
    range_x = lr.real - ul.real
    range_y = ul.imag - lr.imag
    return complex(
        ul.real + float(x) * range_x / float(width),
        ul.imag - float(y) * range_y / float(height)
    )


if __name__ == '__main__':
    args = docopt(__doc__, version='mandel2 0.1.0')

    width = int(args["<width>"])
    height = int(args["<height>"])
    upper_left = complex(args["<upper-left>"])
    lower_right = complex(args["<lower-right>"])

    #
    start = time.time()

    image = np.zeros((height, width), dtype=np.int16)
    for y in xrange(0, height):
        for x in xrange(0, width):
            z = pixel_to_point(width, height, y, x, upper_left, lower_right)
            image[y,x] = 255 - mandel.escapes(z, 255, 255)

    end = time.time()
    elapsed = end - start
    print "C++ inner parts: ", elapsed


    imshow(image)
    show()

    del image

    #
    start = time.time()

    image = np.array(
        mandel.render(width, height, upper_left, lower_right)
    ).reshape(height, width)

    end = time.time()
    elapsed = end - start
    print "C++ for loop:    ", elapsed

    imshow(image)
    show()
